src/xrpld/app/tx/detail/SetTrust.cpp

```diff
@@ -408,15 +413,35 @@ SetTrust::doApply()
             uFlagsOut &= ~(bHigh ? lsfHighNoRipple : lsfLowNoRipple);
         }
 
+        bool willSetFreeze = false;
         if (bSetFreeze && !bClearFreeze && !sle->isFlag(lsfNoFreeze))
         {
             uFlagsOut |= (bHigh ? lsfHighFreeze : lsfLowFreeze);
+            willSetFreeze = true;
         }
         else if (bClearFreeze && !bSetFreeze)
         {
             uFlagsOut &= ~(bHigh ? lsfHighFreeze : lsfLowFreeze);
         }
 
+        bool const alreadyFrozen =
+            bHigh ? sle->isFlag(lsfHighFreeze) : sle->isFlag(lsfLowFreeze);
+        if (bSetDeepFreeze && !bClearDeepFreeze &&
+            (willSetFreeze || alreadyFrozen))
+        {
+            if (ctx_.view().rules().enabled(featureDeepFreeze))
+                uFlagsOut |= (bHigh ? lsfHighDeepFreeze : lsfLowDeepFreeze);
+            else
+                return temMALFORMED;
+        }
+        else if (bClearDeepFreeze && !bSetDeepFreeze)
+        {
+            if (ctx_.view().rules().enabled(featureDeepFreeze))
+                uFlagsOut &= ~(bHigh ? lsfHighDeepFreeze : lsfLowDeepFreeze);
+            else
+                return temMALFORMED;
+        }
```

Comment:
Wouldn't the following be easier to follow?

```suggestion
        if (bSetFreeze && !bClearFreeze && !sle->isFlag(lsfNoFreeze))
        {
            uFlagsOut |= (bHigh ? lsfHighFreeze : lsfLowFreeze);
        }
        else if (bClearFreeze && !bSetFreeze)
        {
            uFlagsOut &= ~(bHigh ? lsfHighFreeze : lsfLowFreeze);
        }

        // First (un)set the deep freeze flag independently of the regular freeze flag.
        if (bSetDeepFreeze && !bClearDeepFreeze && !sle->isFlag(lsfNoFreeze))
        {
                uFlagsOut |= (bHigh ? lsfHighDeepFreeze : lsfLowDeepFreeze);
        }
        else if (bClearDeepFreeze && !bSetDeepFreeze)
        {
                uFlagsOut &= ~(bHigh ? lsfHighDeepFreeze : lsfLowDeepFreeze);
        }
                
        // Next ensure the deep freeze flag can only be set when the regular freeze flag is set.
        if ((!(uFlagsOut & lsfLowFreeze) && (uFlagsOut & lsfLowDeepFreeze)) ||
            (!(uFlagsOut & lsfHighFreeze) && (uFlagsOut & lsfHighDeepFreeze)))
        {
                return temMALFORMED;
        }    
```

If you wish you can gate the deep freeze section by `if (bDeepFreezeEnabled)`, but since the set and clear flags will be zero when deep freeze is disabled, this wouldn't be absolutely necessary. There'd be a tiny performance hit as long as deep freeze is disabled (namely, extra conditional statements that would always be no-ops), while gaining performance when it is enabled (namely, not performing the deep freeze enabled check).

Reply:
I was looking into this and I have a strong feeling that we miss the correct check on ClearFreeze. It should also include checking if the trust line is not deep frozen.